ext/drbbdup/drbbdup.h

```diff
@@ -0,0 +1,334 @@
+                                           void *pre_analysis_data, void *analysis_data);
+
+/***************************************************************************
+ * INIT
+ */
+
+/**
+ * Specifies the options when initialising drbbdup. \p set_up_bb_dups, \p insert_encode
+ * and \p instrument_instr cannot be NULL, while \p dup_limit must be greater than zero.
+ */
+typedef struct {
+    /** Set this to the size of this structure. */
+    size_t struct_size;
+    /** A user-defined call-back function that sets up how to duplicate a basic block.
+     * Cannot be NULL.
+     */
+    drbbdup_set_up_bb_dups_t set_up_bb_dups;
+    /**
+     * A user-defined call-back function that inserts code to encode the runtime case.
+     * The resulting encoding is used by the dispatcher to direct control to the
+     * appropriate basic block. Cannot be NULL.
+     */
+    drbbdup_insert_encode_t insert_encode;
+    /**
+     * A user-defined call-back function that conducts a preliminary analysis on a basic
+     * block.
+     */
+    drbbdup_pre_analyse_bb_t pre_analyse_bb;
+    /**
+     * A user-defined call-back function that destroys preliminary analysis data.
+     */
+    drbbdup_destroy_pre_analysis_t destroy_pre_analysis;
+    /**
+     * A user-defined call-back function that analyses a basic block for a particular
+     * case.
+     */
+    drbbdup_analyse_bb_t analyse_bb;
+    /**
+     * A user-defined call-back function that destroys analysis data for a particular
+     * case.
+     */
+    drbbdup_destroy_analysis_t destroy_analysis;
+    /**
+     * A user-defined call-back function that instruments an instruction with respect to a
+     * particular case.
+     */
+    drbbdup_instrument_instr_t instrument_instr;
+    /* User-data made available to user-defined call-back functions that drbbdup invokes
+     * to manage basic block duplication.
+     */
+    void *user_data;
+    /* The maximum number of cases, excluding the default case, that can be associated
+     * with a basic block. Once the limit is reached and an unhandled case is encountered,
+     * control is directed to the default case.
+     */
+    ushort dup_limit;
+} drbbdup_options_t;
+
+/**
+ * Priorities of drmgr instrumentation passes used by drbbdup. Users
+ * can perform app2app manipulations prior duplication
+ * by ordering such changes before DRMGR_PRIORITY_DRBBDUP.
+ */
+enum {
+    /** Priority of drbbdup. */
+    DRMGR_PRIORITY_DRBBDUP = -1500
+};
+
+/**
+ * Name of drbbdup priorities for analysis and insert steps.
+ */
+#define DRMGR_PRIORITY_NAME_DRBBDUP "drbbdup"
+
+DR_EXPORT
+/**
+ * Initialises the drbbdup extension. Must be called before use of any other routines.
+ *
+ * It cannot be called multiple times as duplication management is specific to a single
+ * use-case where only a one default case encoding is associated with each basic block.
+ *
+ * The \p ops_in parameter are options which dictate how drbbdup manages
```

Comment:
grammar: s/are options/is a set of options/

Reply:
Done.